src/net.md

```diff
@@ -876,6 +877,99 @@ fn run() -> Result<()> {
 # quick_main!(run);
 ```
 
+[ex-check-broken-links]: #ex-check-broken-links
+<a name="ex-check-broken-links"/>
+## Check a webpage for broken links
+
+[![reqwest-badge]][reqwest] [![select-badge]][select] [![url-badge]][url] [![cat-net-badge]][cat-net]
+
+We parse the URL using [`Url::parse`] and then retrieve the document using reqwest.
+
+We call "get_base_url" to retrieve the base URL. If the document has a "base" tag, we get the "href" [`attr`] from the first occurrence of the "base" tag. This is then used as the base URL. Otherwise, we can use [`Position::BeforePath`] with the original URL to get the base of that URL.
+
+We then retrieve all the links from the document, and [`collect`] them into a [`HashSet`] to remove any duplicates.
+
+Now that we have our URL, we pass them into the "get_broken_links" function along with the base URL. This will iterate over the URL, use [`url::ParseOptions`] and [`Url::parse`] for each to get the absolute URL and then perform a HTTP GET request. If the response status is 404, the URL will be added to the HashSet which is returned by the function.
+
+Finally, we check the contents of the HashSet. If it is empty, we report that no links are broken, otherwise we iterate over the HashSet and report the broken links.
+
+```rust,no_run
+# #[macro_use]
+# extern crate error_chain;
+extern crate reqwest;
+extern crate select;
+extern crate url;
+
+use url::{Url, Position};
+use reqwest::StatusCode;
+use select::document::Document;
+use select::predicate::Name;
+
+use std::collections::HashSet;
+#
+# error_chain! {
+#   foreign_links {
+#       ReqError(reqwest::Error);
+#       IoError(std::io::Error);
+#       UrlParseError(url::ParseError);
+#   }
+# }
+fn get_base_url(url: &Url, doc: &Document) -> Result<Url> {
+    let base_tag_href = doc.find(Name("base")).filter_map(|n| n.attr("href")).nth(0);
+
+    let base_url = base_tag_href.map_or_else(|| Url::parse(&url[..Position::BeforePath]), Url::parse)?;
+    
+    Ok(base_url)
+}
+
+fn get_broken_links<'a>(base_url: &Url, links: &HashSet<&'a str>) -> Result<HashSet<&'a str>> {
+    let mut broken_links = HashSet::new();
+    let base_parser = Url::options().base_url(Some(base_url));
+
+    for link in links {
+        let final_url = base_parser.parse(link)?;
+
+        let res = reqwest::get(final_url)?;
+
+        if res.status() == StatusCode::NotFound {
+            broken_links.insert(*link);
+        }
+    }
+
+    Ok(broken_links)
+}
+
+fn run() -> Result<()> {
+    let url = Url::parse("https://www.rust-lang.org/en-US/")?;
+
+    let res = reqwest::get(url.as_ref())?;
+    let document = Document::from_read(res)?;
+
+    let base_url = get_base_url(&url, &document)?;
+
+    let links: HashSet<&str> = document
+        .find(Name("a"))
+        .filter_map(|n| n.attr("href"))
+        .collect();
```

Comment:
How would I do this as `parse` returns a `Result` and therefore I cannot use the "?" operator when iterating? Should I just ignore any URL that cannot be parsed (and report no errors) like this?

    let links: HashSet<Url> = document
        .find(Name("a"))
        .filter_map(|n| n.attr("href"))
        .filter_map(|link| base_parser.parse(link).ok())
        .collect();

Reply:
Well that entirely depends on what you'd like to achieve. Test only the correct links on best effort basis (your suggestion) or error-out even if there is at least one error
by collecting into `Result`
```rust
    let hs: Result<HashSet<_>> =  document
    .find(Name("a"))
    .filter_map(|n| n.attr("href"))
        .map(|link| base_parser.parse(link).map_err(|e| e.into()))
        .collect();
```

Unfortunately to do this trick with error_chain you'll have to jump through hoops with `map_err` as above (normally this would not be required)